### PR TITLE
Alternative Godot 4 port

### PR DESCRIPTION
**Pull request by farfalk**
Godot 4 port, fully functional (afaik) and currently updated to Godot 4 beta 2. Alternative candidate at closing https://github.com/imjp94/gd-YAFSM/issues/46. 

I'm using this addon in a project of mine. At the start I was using the port by @adamscott (https://github.com/imjp94/gd-YAFSM/pull/47) , but the list of corrections I had to make during the development to make it functional (most of all, the real time debugging part) has grown too big to just report them all in the discussion. 

Should be at feature parity, with only a small aesthetic addition: during a transition, the conditions of the transition are highlighted as well. 

Feel free to test this version, merge it in adamscott's proposal, or use it instead!

(thanks a lot to @adamscott for the initial work!)

### CODE REVIEW

**Comment by farfalk** [comment]
I realize my latest addition is a bit off-scope for a "port to Godot 4.0" pull request. The zoom/scaling handling with snapping was very broken with 4.0 though, so I think the changes I made are sensible. Let me know how to proceed otherwise.

**Comment by imjp94** [comment]
It's ok, I think it is absolutely fine to include some minor fix directly into the port.
Thanks for the contribution!

File: addons/imjp94.yafsm/scenes/flowchart/FlowChartGrid.gd [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    for i in range(from.x, from.x + l.x):
+        var color
+
+        if (int(abs(i)) % 10 == 0):
+            color = grid_major
+        else:
+            color = grid_minor
+
+        var base_ofs = i * snap
+        draw_line(Vector2(base_ofs, offset.y), Vector2(base_ofs, corrected_size.y), color, -1, true)
```
Comment (adamscott) [inline addons/imjp94.yafsm/scenes/flowchart/FlowChartGrid.gd:38]: Setting the last parameter (antialiasing) to `false` greatly reduces the perceptible lag without sacrificing quality.

File: addons/imjp94.yafsm/scenes/flowchart/FlowChartGrid.gd [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    for i in range(from.y, from.y + l.y):
+        var color;
+
+        if (int(abs(i)) % 10 == 0):
+            color = grid_major
+        else:
+            color = grid_minor
+
+        var base_ofs = i * snap
+        draw_line(Vector2(offset.x, base_ofs), Vector2(corrected_size.x, base_ofs), color, -1, true)
```
Comment (adamscott) [inline addons/imjp94.yafsm/scenes/flowchart/FlowChartGrid.gd:50]: Setting the last parameter (antialiasing) to false greatly reduces the perceptible lag without sacrificing quality.

**Comment by imjp94** [comment]
Since Godot-4.0 stable has been released, I am preparing to merge 4.0 branch to master as well.
I tested it and found some minor issues:
- [ ] Default state name's index is random. Newly added state should looks like this: State1, State2 and State3
- [ ] Missing default condition name. There should be a default name for newly added condition
- [ ] Nested state shape doesn't revert to normal state shape. Empty nested state is still a hard corner rectangle instead of round corner rectangle. This happens when double click on a state but go back without adding any state into it.

It should be ready to merge after resolving the issues.
Great job!

**Comment by farfalk** [comment]
Thanks for the review! I should be able to work on it in the next few days.

**Comment by farfalk** [comment]
immagine
immagine

@adamscott I had to revert the last parameter of the `draw_line` function to true in `FlowChartGrid.gd` because I was experiencing heavy graphic glitches in my secondary monitor.

**Comment by farfalk** [comment]
@imjp94 should have solved all of your points! Let me know

**Comment by imjp94** [comment]
@farfalk Thank you! It's almost there!

- Can you make the default condition name as "Param" and indexed like the state name? For example, "Param", "Param1", "Param2", etc. 
So it make sense to call `StateMachine.set_param` in code. Sorry that I didn't make it clear in the last comment.
_Godot 3.x(Expected behavior)_
Screenshot 2023-04-28 205048

- And nested state shape doesn't revert to normal state shape.
_Godot 3.x(Expected behavior)_
ezgif com-video-to-gif (1)
_Godot 4.x_
ezgif com-video-to-gif

**Comment by farfalk** [comment]
> @farfalk Thank you! It's almost there!
> 
>     * Can you make the default condition name as "Param" and indexed like the state name? For example, "Param", "Param1", "Param2", etc.
>       So it make sense to call `StateMachine.set_param` in code. Sorry that I didn't make it clear in the last comment.
>       _Godot 3.x(Expected behavior)_
>       Screenshot 2023-04-28 205048
> 
>     * And nested state shape doesn't revert to normal state shape.
>       _Godot 3.x(Expected behavior)_
>       ezgif com-video-to-gif (1)
>             
>               [
>                 
>               
>                   ezgif com-video-to-gif (1)
>                 ](https://user-images.githubusercontent.com/11486079/235154358-f9cd49d2-6444-40b1-8694-2b285a363608.gif)
>               
>               
>                 
>                   
>                     
>                   
>                   
>                     
>                     
>                   
>                 
>                 [
>                   
>                     
>                   
>                 ](https://user-images.githubusercontent.com/11486079/235154358-f9cd49d2-6444-40b1-8694-2b285a363608.gif)
>               
>             
>       _Godot 4.x_
>       ezgif com-video-to-gif
>             
>               [
>                 
>               
>                   ezgif com-video-to-gif
>                 ](https://user-images.githubusercontent.com/11486079/235154373-48217c26-68e3-4dea-a9da-66fac79f6fcd.gif)
>               
>               
>                 
>                   
>                     
>                   
>                   
>                     
>                     
>                   
>                 
>                 [
>                   
>                     
>                   
>                 ](https://user-images.githubusercontent.com/11486079/235154373-48217c26-68e3-4dea-a9da-66fac79f6fcd.gif)

@imjp94 I'll work on the first point asap. For the second, it's a bit strange: on my part it never happens after commit [7715c1c](https://github.com/imjp94/gd-YAFSM/pull/50/commits/7715c1c3c1895ab368737aa0b5572d1273fa2c72). I'll think about it

**Comment by farfalk** [comment]
@imjp94 ... I think I'm encountering some kind of very weird bug. Maybe you know what's happening? In `StateMachineEditor.gd:169`,
this:
```gdscript
if node:
  print_debug("")
  if (not ("states" in node.state) or (node.state.states=={})):
	  print_debug("")
	  # Convert state machine node back to state node
	  convert_to_state(layer, node)
```
is executed. But if I remove the `print_debug("")`s, the code seemingly isn't executed, and the conversion doesn't happen. Am i going crazy?

**Comment by farfalk** [comment]
@imjp94 fixed both, I think. Check it out :)

**Comment by imjp94** [comment]
@farfalk Perfect! Thank you so much!

**Pull request merged**